### ISSUE TITLE
Gossip based webrtc signalling over streams

### ISSUE THREAD

**Issue by mafintosh** [mafintosh/todo#1]
The new hypercore extension api would be great for this

**Comment by hackergrrl** [comment]
Could you say more about your thoughts on this? Very interested.

**Comment by mafintosh** [comment]
yes. basically lets say you have the following topology

```
   a
  / \
 b   c
      \
       d
```
1. peer `c` wants to connect to more peers so it sends a message to `a` with a digest of the peers it's connected to (minus `a` since that's trivial). the digest is `(d)`.
2. since peer `a` has a peer (`b`) that is not in that digest it sends a message to both `c` and `b` that they should forward signalling data (it picks a random one as the initiator).
3. this allows `b` and `c` to connect directly to each other using `a` as a signalling server

**Comment by mafintosh** [comment]
then later one `c` ask `b` for the same thing etc to form a more connected topology

**Comment by hackergrrl** [comment]
This seems really powerful: peers build routing tables to each over time via gossip, and volunteer themselves as relays for signalling data. It could even be agnostic to topology -- peers could share gossip selectively to encourage the formation of topologies that the network prefers. I'm not familiar with hypercore, but this seems like a powerful enough module to be standalone.

I had been thinking along similar lines recently: about a flooding-based approach to add to `webrtc-swarm`, where a fresh node could flood an introduction signal message to the network (with some TTL), gather signal responses from interested parties, and then connect to whoever seemed the most appealing (with the goal of getting a uniform distro of peer distances, probably). Flooding probably won't scale, but it means only one round-trip to get the signal data exchange done. Periodic gossip seems much better for long term network health though.

---

cc @diasdavid, who is very well versed on this subject.

**Comment by mafintosh** [comment]
yup i'm gonna make this into a separate module. my main application for this was to use it in hypercore (a simple distributed append-only log network)

does webrtc support flooding the same signalling data to multiple peers?

**Comment by hackergrrl** [comment]
If only. Each invitation can only be accepted once, it seems. So a node
would need to flood N times for N new peers.

**Comment by lmatteis** [comment]
The only real problem with this is churn: when users go offline and then go back online they need to restart from bootstrap servers (using IP addresses/DNS names). They can't restart via previously discovered webrtc tabs, because they need to be reintroduced and redo the handshaking.
